knowledge/knowledge_read.go

```diff
@@ -23,190 +23,184 @@ import (
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)"
+	}
+	params := map[string]*knowledgepb.InputParam{}
+	resp, err := c.client.IdentityKnowledgeRead(ctx, &knowledgepb.IdentityKnowledgeReadRequest{
+		Query:       query,
+		InputParams: params,
+		Returns:     returns,
 	}, opts...)
 	if err != nil {
 		return nil, err
 	}
-	return parseMultipleNodesFromPaths(resp.GetPaths())
+	return ParseMultipleNodesFromNodes(resp.GetNodes())
 }
 
 // ListNodesByProperty is a helper function that lists all nodes that have the specified type and property.
 func (c *Client) ListNodesByProperty(
 	ctx context.Context,
-	nodeType string,
-	property *knowledgepb.Property,
+	property *knowledgeobjects.Property,
+	returns []*knowledgepb.Return,
+	isIdentity bool,
 	opts ...grpc.CallOption,
-) ([]*knowledgepb.Node, error) {
+) ([]*knowledgeobjects.Node, error) {
 	ctx = insertMetadata(ctx, c.xMetadata)
-	path := fmt.Sprintf("(n:%s)", nodeType)
-	conditions := fmt.Sprintf("WHERE n.%s = $%s", property.Key, property.Key)
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)"
+	}
+	query = fmt.Sprintf(
+		"%s WHERE n.%s=$%s",
+		query,
+		property.Type,
+		property.Type,
+	)
 	params := make(map[string]*knowledgepb.InputParam)
 	switch property.Value.Value.(type) {
 	case *objects.Value_IntegerValue:
-		params[property.Key] = &knowledgepb.InputParam{
+		params[property.Type] = &knowledgepb.InputParam{
 			Value: &knowledgepb.InputParam_IntegerValue{
 				IntegerValue: property.GetValue().GetIntegerValue()}}
 	case *objects.Value_StringValue:
-		params[property.Key] = &knowledgepb.InputParam{
+		params[property.Type] = &knowledgepb.InputParam{
 			Value: &knowledgepb.InputParam_StringValue{
 				StringValue: property.GetValue().GetStringValue()}}
 	default:
 		return nil, errors.New(codes.InvalidArgument, "only string or integer properties can be used for queries")
 	}
-	resp, err := c.client.IdentityKnowledge(ctx, &knowledgepb.IdentityKnowledgeRequest{
-		Path:        path,
-		Conditions:  conditions,
+	resp, err := c.client.IdentityKnowledgeRead(ctx, &knowledgepb.IdentityKnowledgeReadRequest{
+		Query:       query,
 		InputParams: params,
-		Operation:   knowledgepb.Operation_OPERATION_READ,
+		Returns:     returns,
 	}, opts...)
 	if err != nil {
 		return nil, err
 	}
-	return parseMultipleNodesFromPaths(resp.GetPaths())
+	return ParseMultipleNodesFromNodes(resp.GetNodes())
 }
 
-func (c *Client) getNodeByID(
+func (c *Client) GetNodeByID(
 	ctx context.Context,
 	id string,
-	nodeType string,
+	returns []*knowledgepb.Return,
+	isIdentity bool,
 	opts ...grpc.CallOption,
-) (*knowledgepb.Node, error) {
+) (*knowledgeobjects.Node, error) {
 	ctx = insertMetadata(ctx, c.xMetadata)
-	path := fmt.Sprintf("(n:%s)", nodeType)
-	conditions := fmt.Sprintf("WHERE n.%s = $%s", ID, ID)
+	query := "MATCH (n:Resource)"
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)"
+	}
+	query = fmt.Sprintf(
+		"%s WHERE n.%s=$%s",
+		query,
+		ID,
+		ID,
+	)
```

Comment:
But this is fine, since the `id` is stored on the node, not as a separate property node. Same with `external_id`